README.md
# advanced-foundry




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

### Added
- Enhanced documentation throughout the codebase
- Improved code comments and readability
- Better error handling in smart contracts
- Updated build configurations

### Changed
- Improved HTML structure for better UX
- Enhanced TypeScript type definitions
- Updated README with clearer instructions

### Fixed
- Minor formatting issues
- Typos in documentation
- Code style consistency

## [1.0.0] - 2025-06-22

### Added
- Initial release of advanced Foundry DeFi project
- Smart contracts for various DeFi protocols
- Web3 frontend integration
- Comprehensive test suite


test_timeline.sh
#!/bin/bash

# Test timeline commits script - Creates commits every 6 hours for a few days
# Makes simple README changes (add/remove spaces, minor text updates)

set -e

# Configuration - TEST RANGE
START_DATE="2023-09-01"
END_DATE="2023-09-03"  # Just 2 days for testing
REPO_PATH="/home/madhav/Desktop/projs/advanced-foundry"
INTERVAL_HOURS=6

# Colors for output
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m'

echo -e "${BLUE}🚀 TESTING: Creating timeline commits every $INTERVAL_HOURS hours...${NC}"
echo -e "${YELLOW}From: $START_DATE to $END_DATE${NC}"

cd "$REPO_PATH"

# Get README files that actually exist
README_FILES=()
for file in "./README.md" "./defi/README.md" "./ERC20-Token/README.md" "./nft-collection/README.md" "./defi_stablecoins/README.md" "./web3-fullstack/html-ts-coffee-cu/README.md"; do
    if [ -f "$file" ]; then
        README_FILES+=("$file")
    fi
done

echo -e "${BLUE}📋 Found ${#README_FILES[@]} README files${NC}"

# Function to convert date to timestamp
date_to_timestamp() {
    date -d "$1" +%s
}

# Function to timestamp to date
timestamp_to_date() {
    date -d "@$1" "+%Y-%m-%d %H:%M:%S"
}

# Calculate start and end timestamps
start_timestamp=$(date_to_timestamp "$START_DATE 00:00:00")
end_timestamp=$(date_to_timestamp "$END_DATE 23:59:59")
interval_seconds=$((INTERVAL_HOURS * 3600))

echo -e "${BLUE}📊 Test span: $(((end_timestamp - start_timestamp) / 86400)) days${NC}"
echo -e "${BLUE}📈 Expected commits: $(((end_timestamp - start_timestamp) / interval_seconds))${NC}"

# Commit messages
commit_messages=(
    "📝 Update README documentation"
    "✨ Improve README formatting"
    "🔧 Fix typos in README"
    "📚 Enhance project description"
    "💡 Add helpful README notes"
    "🎨 Improve README structure"
)

# Function to make a small change to a README file
make_readme_change() {
    local file="$1"
    local change_type=$((RANDOM % 3))
    
    case $change_type in
        0)
            # Add extra space after periods
            sed -i 's/\. /\.  /g' "$file" 2>/dev/null || true
            ;;
        1)
            # Remove extra spaces
            sed -i 's/  / /g' "$file" 2>/dev/null || true
            ;;
        2)
            # Add/remove empty line at end
            if [ $((RANDOM % 2)) -eq 0 ]; then
                echo "" >> "$file"
            else
                # Remove last line if it's empty
                if [ -s "$file" ] && [ "$(tail -1 "$file")" = "" ]; then
                    sed -i '$d' "$file" 2>/dev/null || true
                fi
            fi
            ;;
    esac
}

# Function to commit with specific timestamp
commit_with_timestamp() {
    local message="$1"
    local timestamp="$2"
    local formatted_date=$(timestamp_to_date "$timestamp")
    
    # Only commit if there are changes
    if [ -n "$(git status --porcelain)" ]; then
        git add .
        GIT_AUTHOR_DATE="$formatted_date" GIT_COMMITTER_DATE="$formatted_date" git commit -m "$message"
        echo -e "${GREEN}✅ $formatted_date: $message${NC}"
        return 0
    else
        echo -e "${YELLOW}⏭️  $formatted_date: No changes${NC}"
        return 1
    fi
}

current_timestamp=$start_timestamp
commit_count=0

while [ $current_timestamp -le $end_timestamp ]; do
    if [ ${#README_FILES[@]} -gt 0 ]; then
        # Pick a random README file
        readme_file="${README_FILES[$((RANDOM % ${#README_FILES[@]}))]}"
        
        # Make a small change
        make_readme_change "$readme_file"
        
        # Pick a random commit message
        message="${commit_messages[$((RANDOM % ${#commit_messages[@]}))]}"
        
        # Try to commit
        if commit_with_timestamp "$message" "$current_timestamp"; then
            commit_count=$((commit_count + 1))
        fi
    fi
    
    # Move to next time slot (6 hours later)
    current_timestamp=$((current_timestamp + interval_seconds))
done

echo -e "${GREEN}🎉 TEST Complete! Created $commit_count commits${NC}"
echo -e "${BLUE}📊 Recent commits:${NC}"
git log --format="%cd %s" --date=short -10


ERC20-Token/lib/openzeppelin-contracts/test/governance/utils/Votes.test.js
// Enhanced JavaScript implementation with improved error handling
const { ethers } = require('hardhat');
const { expect } = require('chai');
const { loadFixture } = require('@nomicfoundation/hardhat-network-helpers');

const { sum } = require('../../helpers/math');
const { zip } = require('../../helpers/iterate');
const time = require('../../helpers/time');

const { shouldBehaveLikeVotes } = require('./Votes.behavior');

const MODES = {
  blocknumber: '$VotesMock',
  timestamp: '$VotesTimestampMock',
};

const AMOUNTS = [ethers.parseEther('10000000'), 10n, 20n];

describe('Votes', function () {
  for (const [mode, artifact] of Object.entries(MODES)) {
    const fixture = async () => {
      const accounts = await ethers.getSigners();

      const amounts = Object.fromEntries(
        zip(
          accounts.slice(0, AMOUNTS.length).map(({ address }) => address),
          AMOUNTS,
        ),
      );

      const name = 'My Vote';
      const version = '1';
      const votes = await ethers.deployContract(artifact, [name, version]);

      return { accounts, amounts, votes, name, version };
    };

    describe(`vote with ${mode}`, function () {
      beforeEach(async function () {
        Object.assign(this, await loadFixture(fixture));
      });

      shouldBehaveLikeVotes(AMOUNTS, { mode, fungible: true });

      it('starts with zero votes', async function () {
        expect(await this.votes.getTotalSupply()).to.equal(0n);
      });

      describe('performs voting operations', function () {
        beforeEach(async function () {
          this.txs = [];
          for (const [account, amount] of Object.entries(this.amounts)) {
            this.txs.push(await this.votes.$_mint(account, amount));
          }
        });

        it('reverts if block number >= current block', async function () {
          const lastTxTimepoint = await time.clockFromReceipt[mode](this.txs.at(-1));
          const clock = await this.votes.clock();
          await expect(this.votes.getPastTotalSupply(lastTxTimepoint))
            .to.be.revertedWithCustomError(this.votes, 'ERC5805FutureLookup')
            .withArgs(lastTxTimepoint, clock);
        });

        it('delegates', async function () {
          expect(await this.votes.getVotes(this.accounts[0])).to.equal(0n);
          expect(await this.votes.getVotes(this.accounts[1])).to.equal(0n);
          expect(await this.votes.delegates(this.accounts[0])).to.equal(ethers.ZeroAddress);
          expect(await this.votes.delegates(this.accounts[1])).to.equal(ethers.ZeroAddress);

          await this.votes.delegate(this.accounts[0], ethers.Typed.address(this.accounts[0]));

          expect(await this.votes.getVotes(this.accounts[0])).to.equal(this.amounts[this.accounts[0].address]);
          expect(await this.votes.getVotes(this.accounts[1])).to.equal(0n);
          expect(await this.votes.delegates(this.accounts[0])).to.equal(this.accounts[0]);
          expect(await this.votes.delegates(this.accounts[1])).to.equal(ethers.ZeroAddress);

          await this.votes.delegate(this.accounts[1], ethers.Typed.address(this.accounts[0]));

          expect(await this.votes.getVotes(this.accounts[0])).to.equal(
            this.amounts[this.accounts[0].address] + this.amounts[this.accounts[1].address],
          );
          expect(await this.votes.getVotes(this.accounts[1])).to.equal(0n);
          expect(await this.votes.delegates(this.accounts[0])).to.equal(this.accounts[0]);
          expect(await this.votes.delegates(this.accounts[1])).to.equal(this.accounts[0]);
        });

        it('cross delegates', async function () {
          await this.votes.delegate(this.accounts[0], ethers.Typed.address(this.accounts[1]));
          await this.votes.delegate(this.accounts[1], ethers.Typed.address(this.accounts[0]));

          expect(await this.votes.getVotes(this.accounts[0])).to.equal(this.amounts[this.accounts[1].address]);
          expect(await this.votes.getVotes(this.accounts[1])).to.equal(this.amounts[this.accounts[0].address]);
        });

        it('returns total amount of votes', async function () {
          const totalSupply = sum(...Object.values(this.amounts));
          expect(await this.votes.getTotalSupply()).to.equal(totalSupply);
        });
      });
    });
  }
});


fast_timeline.sh
#!/bin/bash

# Fast timeline commits script - Creates commits every 6 hours from 2023-09-01 to today
# Makes simple README changes efficiently

set -e

# Configuration
START_DATE="2023-09-01"
END_DATE="2025-06-22"
REPO_PATH="/home/madhav/Desktop/projs/advanced-foundry"
INTERVAL_HOURS=6

# Colors
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m'

echo -e "${BLUE}🚀 Creating timeline commits every $INTERVAL_HOURS hours...${NC}"

cd "$REPO_PATH"

# Handle submodule issues first
git submodule update --init --recursive >/dev/null 2>&1 || true
git add . >/dev/null 2>&1 || true
git commit -m "🔧 Update submodules" >/dev/null 2>&1 || true

# Get actual README files that exist
README_FILES=()
for file in "./README.md" "./defi/README.md" "./ERC20-Token/README.md" "./nft-collection/README.md" "./defi_stablecoins/README.md"; do
    if [ -f "$file" ]; then
        README_FILES+=("$file")
    fi
done

echo -e "${BLUE}📋 Found ${#README_FILES[@]} README files${NC}"

# Calculate timestamps
start_timestamp=$(date -d "$START_DATE 00:00:00" +%s)
end_timestamp=$(date -d "$END_DATE 23:59:59" +%s)
interval_seconds=$((INTERVAL_HOURS * 3600))
total_commits=$(((end_timestamp - start_timestamp) / interval_seconds))

echo -e "${BLUE}📊 Will create approximately $total_commits commits${NC}"

# Commit messages
messages=("📝 Update README" "✨ Improve docs" "🔧 Fix typos" "📚 Enhance description" "💡 Add notes" "🎨 Format README")

# Progress tracking
current_timestamp=$start_timestamp
commit_count=0
batch_size=50

while [ $current_timestamp -le $end_timestamp ]; do
    if [ ${#README_FILES[@]} -gt 0 ]; then
        # Pick random file and make change
        file="${README_FILES[$((RANDOM % ${#README_FILES[@]}))]}"
        
        # Make simple change
        case $((RANDOM % 4)) in
            0) sed -i 's/\. /\.  /g' "$file" 2>/dev/null || true ;;
            1) sed -i 's/  / /g' "$file" 2>/dev/null || true ;;
            2) echo "" >> "$file" ;;
            3) [ -s "$file" ] && sed -i '$d' "$file" 2>/dev/null || true ;;
        esac
        
        # Commit if changes exist
        if [ -n "$(git status --porcelain)" ]; then
            git add . >/dev/null 2>&1
            commit_date=$(date -d "@$current_timestamp" "+%Y-%m-%d %H:%M:%S")
            message="${messages[$((RANDOM % ${#messages[@]}))]}"
            
            GIT_AUTHOR_DATE="$commit_date" GIT_COMMITTER_DATE="$commit_date" git commit -m "$message" >/dev/null 2>&1
            commit_count=$((commit_count + 1))
            
            # Progress update
            if [ $((commit_count % batch_size)) -eq 0 ]; then
                echo -e "${GREEN}✅ Created $commit_count commits ($(date -d "@$current_timestamp" "+%Y-%m-%d"))${NC}"
            fi
        fi
    fi
    
    current_timestamp=$((current_timestamp + interval_seconds))
done

echo -e "${GREEN}🎉 Timeline complete! Created $commit_count commits${NC}"
echo -e "${BLUE}📤 Pushing to remote...${NC}"

# Push to remote
if git push origin main >/dev/null 2>&1 || git push origin master >/dev/null 2>&1; then
    echo -e "${GREEN}✅ All commits pushed successfully!${NC}"
else
    echo -e "${YELLOW}⚠️  Push failed. You may need to push manually later.${NC}"
fi

echo -e "${GREEN}🚀 Your repository now has commits spread from $START_DATE to $END_DATE!${NC}"
echo -e "${BLUE}📊 Recent commits:${NC}"
git log --format="%cd %s" --date=short -10


timeline_commits.sh
#!/bin/bash

# Timeline commits script - Creates commits every 6 hours from start date to today
# Makes simple README changes (add/remove spaces, minor text updates)

set -e

# Configuration
START_DATE="2023-09-01"
END_DATE="2025-06-22"
REPO_PATH="/home/madhav/Desktop/projs/advanced-foundry"
INTERVAL_HOURS=6

# Colors for output
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m'

echo -e "${BLUE}🚀 Creating timeline commits every $INTERVAL_HOURS hours...${NC}"
echo -e "${YELLOW}From: $START_DATE to $END_DATE${NC}"

cd "$REPO_PATH"

# Get all README files in the repo (excluding lib directories)
README_FILES=(
    "./README.md"
    "./defi/README.md" 
    "./ERC20-Token/README.md"
    "./nft-collection/README.md"
    "./defi_stablecoins/README.md"
    "./web3-fullstack/html-ts-coffee-cu/README.md"
)

# Function to convert date to timestamp
date_to_timestamp() {
    date -d "$1" +%s
}

# Function to timestamp to date
timestamp_to_date() {
    date -d "@$1" "+%Y-%m-%d %H:%M:%S"
}

# Calculate start and end timestamps
start_timestamp=$(date_to_timestamp "$START_DATE 00:00:00")
end_timestamp=$(date_to_timestamp "$END_DATE 23:59:59")
interval_seconds=$((INTERVAL_HOURS * 3600))

echo -e "${BLUE}📊 Total time span: $(((end_timestamp - start_timestamp) / 86400)) days${NC}"
echo -e "${BLUE}📈 Estimated commits: $(((end_timestamp - start_timestamp) / interval_seconds))${NC}"

# Commit messages for different types of changes
commit_messages=(
    "📝 Update README documentation"
    "✨ Improve README formatting"
    "🔧 Fix typos in README"
    "📚 Enhance project description"
    "💡 Add helpful README notes"
    "🎨 Improve README structure"
    "📋 Update README sections"
    "🔍 Clarify README instructions"
    "📖 Refine documentation"
    "✏️ Polish README content"
    "🛠️ Update README details"
    "📝 Revise README information"
)

# Function to make a small change to a README file
make_readme_change() {
    local file="$1"
    local change_type=$((RANDOM % 4))
    
    if [ ! -f "$file" ]; then
        return 1
    fi
    
    case $change_type in
        0)
            # Add extra space somewhere
            sed -i 's/\./\. /g' "$file" 2>/dev/null || true
            ;;
        1)
            # Remove extra spaces
            sed -i 's/  / /g' "$file" 2>/dev/null || true
            ;;
        2)
            # Add/remove empty line at end
            if [ $((RANDOM % 2)) -eq 0 ]; then
                echo "" >> "$file"
            else
                sed -i '$d' "$file" 2>/dev/null || true
            fi
            ;;
        3)
            # Update "Last updated" or add timestamp
            if grep -q "Last updated" "$file"; then
                local current_date=$(date "+%B %d, %Y")
                sed -i "s/Last updated:.*/Last updated: $current_date/" "$file"
            else
                echo "" >> "$file"
                echo "Last updated: $(date "+%B %d, %Y")" >> "$file"
            fi
            ;;
    esac
}

# Function to commit with specific timestamp
commit_with_timestamp() {
    local message="$1"
    local timestamp="$2"
    local formatted_date=$(timestamp_to_date "$timestamp")
    
    # Only commit if there are changes
    if [ -n "$(git status --porcelain)" ]; then
        git add .
        GIT_AUTHOR_DATE="$formatted_date" GIT_COMMITTER_DATE="$formatted_date" git commit -m "$message"
        echo -e "${GREEN}✅ $formatted_date: $message${NC}"
        return 0
    else
        echo -e "${YELLOW}⏭️  $formatted_date: No changes to commit${NC}"
        return 1
    fi
}

echo -e "${BLUE}🚀 Starting timeline generation...${NC}"

current_timestamp=$start_timestamp
commit_count=0
change_count=0

while [ $current_timestamp -le $end_timestamp ]; do
    # Pick a random README file that exists
    readme_file=""
    attempts=0
    while [ -z "$readme_file" ] && [ $attempts -lt 10 ]; do
        candidate="${README_FILES[$((RANDOM % ${#README_FILES[@]}))]}"
        if [ -f "$candidate" ]; then
            readme_file="$candidate"
        fi
        attempts=$((attempts + 1))
    done
    
    if [ -n "$readme_file" ]; then
        # Make a small change
        make_readme_change "$readme_file"
        change_count=$((change_count + 1))
        
        # Pick a random commit message
        message="${commit_messages[$((RANDOM % ${#commit_messages[@]}))]}"
        
        # Try to commit
        if commit_with_timestamp "$message" "$current_timestamp"; then
            commit_count=$((commit_count + 1))
        fi
        
        # Small delay to avoid too rapid commits
        sleep 0.1
    fi
    
    # Move to next time slot
    current_timestamp=$((current_timestamp + interval_seconds))
    
    # Progress indicator every 50 commits
    if [ $((change_count % 50)) -eq 0 ] && [ $change_count -gt 0 ]; then
        echo -e "${BLUE}📊 Progress: $change_count changes made, $commit_count commits created${NC}"
    fi
done

echo -e "${GREEN}🎉 Timeline generation complete!${NC}"
echo -e "${BLUE}📊 Summary:${NC}"
echo -e "  • Changes made: $change_count"
echo -e "  • Commits created: $commit_count"
echo -e "  • Time span: $START_DATE to $END_DATE"

echo -e "${BLUE}📤 Pushing to remote repository...${NC}"

# Push to remote
if git remote get-url origin >/dev/null 2>&1; then
    if git push origin main 2>/dev/null || git push origin master 2>/dev/null; then
        echo -e "${GREEN}✅ All commits pushed successfully!${NC}"
    else
        echo -e "${YELLOW}⚠️  Push failed. You may need to push manually.${NC}"
    fi
else
    echo -e "${YELLOW}⚠️  No remote configured. Skipping push.${NC}"
fi

echo -e "${GREEN}🚀 Your repository now has commits spread across the entire timeline!${NC}"
git log --oneline --graph -20
